### ISSUE TITLE
N816 fails to find misnamed variables in generator/list comprehensions

### ISSUE THREAD

**Issue by ember91** [astral-sh/ruff#13325]
## List of keywords searched for

 - N816

## Minimal code snippet

```python
all(someType not in [] for someType in [])
all([someType not in [] for someType in []])
someType = 8
```

## Invoked command

This assumes the file is named `test.py`.

```bash
ruff check --isolated --select N816 test.py
```

## Current Ruff settings

-

## Current Ruff version

```
`0.5.5`
```

## Expectations

I expected all of the lines 1-3 to report error N816, but only line 3 did.

**Comment by dhruvmanila** [comment]
Thanks for the report. The rule `N816` is only for the global variable names while in your example the `someVar1` and `someVar2` variables are in the comprehension. So, only the variables that are in the global scope are checked while a comprehension creates it's own scope.

**Comment by ember91** [comment]
Ok. That's my mistake. But still with `ruff check --isolated --select N test.py` there is no check for this case. Seems like something's missing here.

**Comment by ember91** [comment]
I tried this with:
- `flake8==7.1.1`
- `pep8-naming==0.14.1`

and got:
```
test.py:1:21: N816 variable 'someVar1' in global scope should not be mixedCase
test.py:2:21: N816 variable 'someVar2' in global scope should not be mixedCase
test.py:3:2: N816 variable 'someVar3' in global scope should not be mixedCase
```

**Comment by dhruvmanila** [comment]
> That's my mistake. But still with `ruff check --isolated --select N test.py` there is no check for this case. Seems like something's missing here.

Yeah, I don't think a check like that exists which could mainly be a side effect of having a 1-1 implementation with the original plugin in terms of rules.



> and got:
> 
> ```
> test.py:1:21: N816 variable 'someVar1' in global scope should not be mixedCase
> test.py:2:21: N816 variable 'someVar2' in global scope should not be mixedCase
> test.py:3:2: N816 variable 'someVar3' in global scope should not be mixedCase
> ```
> 
> Although I do understand that `someVar1` and `someVar2` most likely are not globals.

Interesting, that seems like a bug because, as you mentioned, those variables are not in the global scope. I can't really find an issue in the issue tracker of `flake8-naming` so I'm not sure why is this the case.

**Comment by dhruvmanila** [comment]
I think I'd be interested to know why is `pep8-naming` raising a violation for this scenario when the variable is not in the global scope. There do exist rules for other scopes like [function](https://docs.astral.sh/ruff/rules/non-lowercase-variable-in-function/) and [class](https://docs.astral.sh/ruff/rules/mixed-case-variable-in-class-scope/). Ruff's semantic model is aware of the fact that comprehensions create a new scope and it could be that `pep8-naming` isn't considering that.

**Comment by ember91** [comment]
I created this one to discuss it a bit: https://github.com/PyCQA/pep8-naming/issues/233
